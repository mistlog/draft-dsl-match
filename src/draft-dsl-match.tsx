import {
    isTSUnionType,
    expressionStatement,
    blockStatement,
    Identifier,
    TypeAnnotation,
    isTSLiteralType,
    IfStatement,
    BlockStatement,
    StringLiteral,
    NumberLiteral,
    isTSTypeReference,
    ArrowFunctionExpression,
    ExpressionStatement,
    isStringLiteral,
    isTSQualifiedName,
    isIdentifier,
} from "@babel/types";
import { ToAst, ToString } from "typedraft";

export class PatternMatch {
    m_Merge: boolean;

    constructor(merge = false) {
        this.m_Merge = merge;
    }
}

//@ts-ignore
<PatternMatch /> +
    function Transcribe(block: Array<ExpressionStatement>): [IfStatement] {
        let tail: IfStatement = null;
        let head: IfStatement = null;

        // if we are in inline context, remove "use match"
        const [first] = block;
        if (first && isStringLiteral(first.expression)) {
            block = block.slice(1);
        }

        block.forEach(each => {
            const expression = each.expression as ArrowFunctionExpression;

            //@ts-ignore
            <TranscribeExpressionToIf />;
        });

        /**
         * transcribed is the "head"(first) of if statement
         */
        return [head];
    };

function TranscribeExpressionToIf(
    expression: ArrowFunctionExpression,
    head: IfStatement,
    tail: IfStatement
) {
    if (expression.params.length === 0) {
        //@ts-ignore
        <HandleDefaultCase />;
    } else {
        //@ts-ignore
        <BuildCurrentIf />;

        //@ts-ignore
        <MoveToNext />;
    }
}

type Literal = (StringLiteral | NumberLiteral) & { extra: { raw: string } };

function BuildCurrentIf(expression: ArrowFunctionExpression) {
    //
    const [pattern_info] = expression.params as [Identifier];
    const to_match = pattern_info.name;
    const annotation = (pattern_info.typeAnnotation as TypeAnnotation).typeAnnotation;

    //
    let current: IfStatement = null;

    if (isTSLiteralType(annotation)) {
        /**
         * number or string
         */
        const pattern = (annotation.literal as Literal).extra.raw;
        current = ToAst(`if(${to_match}===${pattern}){}`) as IfStatement;
    } else if (isTSTypeReference(annotation)) {
        /**
         * enum, instanceof
         */

        let operator = null;
        if (isTSQualifiedName(annotation.typeName)) {
            operator = "===";
        } else if (isIdentifier(annotation.typeName)) {
            operator = "instanceof";
        }
        current = ToAst(`if(${to_match} ${operator} ${ToString(annotation)}){}`) as IfStatement;
    } else if (isTSUnionType(annotation)) {
        /**
         * or
         */
        const types = annotation.types;
        const patterns = types.map(each => {
            if (each.type === "TSLiteralType") {
                return (each.literal as Literal).extra.raw;
            } else if (each.type === "TSTypeReference") {
                return ToString(each);
            }
        });
        const fragments = patterns.map(each => `${to_match}===${each}`);
        const condition = fragments.join("||");
        current = ToAst(`if(${condition}){}`) as IfStatement;
    }

    current.consequent =
        expression.body.type === "BlockStatement"
            ? (expression.body as BlockStatement)
            : blockStatement([expressionStatement(expression.body)]);
}

function MoveToNext(head: IfStatement, current: IfStatement, tail: IfStatement) {
    if (head == null) {
        head = tail = current;
    } else {
        tail = tail.alternate = current;
    }
}

function HandleDefaultCase(expression: ArrowFunctionExpression, tail: IfStatement) {
    tail.alternate =
        expression.body.type === "BlockStatement"
            ? (expression.body as BlockStatement)
            : blockStatement([expressionStatement(expression.body)]);
}
