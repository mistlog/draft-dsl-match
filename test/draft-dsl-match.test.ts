import { LocalContext, ToString, ToBinding, InlineContext, ToNodePath, ToFile } from "typedraft";
import { BlockStatement, Program } from "@babel/types";
import traverse, { NodePath } from "@babel/traverse";
import { PatternMatch } from "../src/draft-dsl-match";

describe("test dsl match", () => {
    test("dsl.match.number", () => {
        //
        const code = `
            function Test(value:number){
                'use match';

                (value: 1) =>
                {
                    console.log(1);
                }
            
                (value: 2) =>
                {
                    console.log(2);
                }
            
                (value: 3) =>
                {
                    console.log(3);
                }
            }
        `;

        SnapshotTest(code);

    })

    test("dsl.match.string", () => {
        //
        const code = `
            function Test(value:string){
                'use match';

                (value: "a") =>
                {
                    console.log("a");
                }
            
                (value: "b") =>
                {
                    console.log("b");
                }
            }
        `;

        SnapshotTest(code);


    })

    test("dsl.match.enum", () => {
        //
        const code = `
            function Test(value:Event){
                'use match';

                (value: Event.EventA) =>
                {
                    console.log(1);
                }
            
                (value: Event.EventB) =>
                {
                    console.log("a");
                }
            }
        `;

        SnapshotTest(code);


    })

    test("dsl.match.instanceof", () => {
        //
        const code = `
            function Test(value:ClassA | ClassB){
                'use match';

                (value: ClassA) =>
                {
                    console.log("class A");
                }
            
                (value: ClassB) =>
                {
                    console.log("class B");
                }
            }
        `;

        SnapshotTest(code);
    })

    test("dsl.match.only-default: only default is not allowed", () => {
        //
        const code = `
            function Test(value:number){
                'use match';

                () =>
                {
                    console.log(2);
                }
            }
        `;

        const context = new LocalContext(ToBinding(code));
        expect(() => context.Resolve(new PatternMatch())).toThrowError("Cannot set property 'alternate' of null");
    })

    test("dsl.match.default", () => {
        //
        const code = `
            function Test(value:number){
                'use match';

                (value: 1) =>
                {
                    console.log(1);
                }
            
                () =>
                {
                    console.log(2);
                }
            }
        `;

        SnapshotTest(code);

    })

    test("dsl.match: return block statement", () => {
        //
        const code = `
            function Test(value:number){
                'use match';

                (value: 1) => console.log(1);
            
                () => <HandleSomething/>;
            }
        `;

        SnapshotTest(code);

    })

    test("dsl.match.or", () => {
        //
        const code = `
            function Test(value:any){
                'use match';

                (value: "a" | "b") =>
                {
                    console.log("string");
                }

                (value: 1 | 2) =>
                {
                    console.log("number");
                }

                (value: Event.A | Event.B) => 
                {
                    console.log("enum");
                }
            }
        `;

        SnapshotTest(code);

    })

    test("dsl.match.inline-context", () => {
        const code = `
            {
                'use match';

                (value: "a" | "b") => {
                    console.log("string");
                }

                (value: 1 | 2) => {
                    console.log("number");
                }
            }
        `;

        const context = new InlineContext(ToNodePath(code));
        context.Resolve(new PatternMatch());
        expect(ToString(context.m_Code)).toMatchSnapshot();
    })

    test("dsl.match.inline-context: merge", () => {
        const code = `
            {
                'use match';

                (value: "a" | "b") => {
                    console.log("string");
                }

                (value: 1 | 2) => {
                    console.log("number");
                }
            }
        `;


        // ToNodePath returns path without container, key, etc... 
        // however, they will be used in merge(path.replaceWithMultiple)
        // so we will get real path by traverse:
        let program_path: NodePath<Program> = null;
        traverse(ToFile(code), { Program(path) { program_path = path; } });
        const [path] = program_path.get("body") as [NodePath<BlockStatement>];

        const context = new InlineContext(path);
        context.Resolve(new PatternMatch(true));

        // path will be invalid after merge, that's fine in typedraft due to RefreshDraftPlugin
        // so we use program_path.node to check the transcribed code:
        expect(ToString(program_path.node)).toMatchSnapshot();
    })
});

function SnapshotTest(code: string) {
    const context = new LocalContext(ToBinding(code));
    context.Resolve(new PatternMatch());
    expect(ToString(context.m_Code)).toMatchSnapshot();
}