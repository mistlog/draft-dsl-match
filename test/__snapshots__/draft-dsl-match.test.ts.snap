// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test dsl match dsl.match.default 1`] = `
"function Test(value: number) {
  if (value === 1) {
    console.log(1);
  } else {
    console.log(2);
  }
}"
`;

exports[`test dsl match dsl.match.enum 1`] = `
"function Test(value: Event) {
  if (value === Event.EventA) {
    console.log(1);
  } else if (value === Event.EventB) {
    console.log(\\"a\\");
  }
}"
`;

exports[`test dsl match dsl.match.inline-context 1`] = `
"{
  if (value === \\"a\\" || value === \\"b\\") {
    console.log(\\"string\\");
  } else if (value === 1 || value === 2) {
    console.log(\\"number\\");
  }
}"
`;

exports[`test dsl match dsl.match.inline-context: merge 1`] = `
"if (value === \\"a\\" || value === \\"b\\") {
  console.log(\\"string\\");
} else if (value === 1 || value === 2) {
  console.log(\\"number\\");
}"
`;

exports[`test dsl match dsl.match.instanceof 1`] = `
"function Test(value: ClassA | ClassB) {
  if (value instanceof ClassA) {
    console.log(\\"class A\\");
  } else if (value instanceof ClassB) {
    console.log(\\"class B\\");
  }
}"
`;

exports[`test dsl match dsl.match.number 1`] = `
"function Test(value: number) {
  if (value === 1) {
    console.log(1);
  } else if (value === 2) {
    console.log(2);
  } else if (value === 3) {
    console.log(3);
  }
}"
`;

exports[`test dsl match dsl.match.or 1`] = `
"function Test(value: any) {
  if (value === \\"a\\" || value === \\"b\\") {
    console.log(\\"string\\");
  } else if (value === 1 || value === 2) {
    console.log(\\"number\\");
  } else if (value === Event.A || value === Event.B) {
    console.log(\\"enum\\");
  }
}"
`;

exports[`test dsl match dsl.match.string 1`] = `
"function Test(value: string) {
  if (value === \\"a\\") {
    console.log(\\"a\\");
  } else if (value === \\"b\\") {
    console.log(\\"b\\");
  }
}"
`;

exports[`test dsl match dsl.match: return block statement 1`] = `
"function Test(value: number) {
  if (value === 1) {
    console.log(1);
  } else {
    <HandleSomething />;
  }
}"
`;
